src/ucp/rndv/rndv_put.c

```diff
@@ -310,3 +358,115 @@ static ucp_proto_t ucp_rndv_put_zcopy_proto = {
     },
 };
 UCP_PROTO_REGISTER(&ucp_rndv_put_zcopy_proto);
+
+
+static void ucp_proto_rndv_put_mtype_pack_completion(uct_completion_t *uct_comp)
+{
+    ucp_request_t *req = ucs_container_of(uct_comp, ucp_request_t,
+                                          send.state.uct_comp);
+    const ucp_proto_rndv_put_priv_t *rpriv = req->send.proto_config->priv;
+
+    ucp_trace_req(req, "mtype_pack_completion mdesc %p", req->send.rndv.mdesc);
+    ucp_proto_completion_init(&req->send.state.uct_comp, rpriv->put_comp_cb);
+    ucp_request_send(req);
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t ucp_proto_rndv_put_mtype_send_func(
+        ucp_request_t *req, const ucp_proto_multi_lane_priv_t *lpriv,
+        ucp_datatype_iter_t *next_iter)
+{
+    const ucp_proto_rndv_put_priv_t *rpriv = req->send.proto_config->priv;
+    uct_iov_t iov;
+
+    ucp_proto_rndv_mtype_next_iov(req, &rpriv->bulk, lpriv, next_iter, &iov);
+    return ucp_proto_rndv_put_common_send(req, lpriv, &iov,
+                                          &req->send.state.uct_comp);
+}
+
+static ucs_status_t
+ucp_proto_rndv_put_mtype_send_progress(uct_pending_req_t *uct_req)
+{
+    ucp_request_t *req = ucs_container_of(uct_req, ucp_request_t, send.uct);
+    const ucp_proto_rndv_put_priv_t *rpriv = req->send.proto_config->priv;
+    ucs_status_t status;
+
+    if (!(req->flags & UCP_REQUEST_FLAG_PROTO_INITIALIZED)) {
+        status = ucp_proto_rndv_mtype_request_init(req);
+        if (status != UCS_OK) {
+            ucp_proto_request_abort(req, status);
+            return UCS_OK;
+        }
+
+        ucp_proto_rndv_put_common_request_init(req);
+        req->flags |= UCP_REQUEST_FLAG_PROTO_INITIALIZED;
+
+        ucp_proto_rndv_mtype_copy(req, uct_ep_get_zcopy,
+                                  ucp_proto_rndv_put_mtype_pack_completion,
+                                  "in from");
+        return UCS_OK;
```

Comment:
why do we return here?

Reply:
we started the memory type copy operation and will resume sending when this copy is calling its completion callback